### PR TITLE
Adds user retirement sink

### PR DESCRIPTION
**Pull request by pomegranited**
**Description:**

Adds an event sink for "user retirement" events.

On receipt of a `USER_RETIRE_LMS_MISC` Django signal for a given user, the sink will delete all records from the PII tables for that user, currently:

* event_sink.user_profile
* event_sink.external_id

**ISSUE:** https://github.com/openedx/tutor-contrib-aspects/issues/449

**Relates to:** https://github.com/openedx/tutor-contrib-aspects/pull/529

This PR should be merged/tagged first.

**Merge deadline:** none

**Installation instructions:**

To install this temporary branch on your local Tutor Aspects stack:

1. Install the Aspects plugin from https://github.com/openedx/tutor-contrib-aspects/pull/529
1. Install the user retirement plugin, following instructions on https://github.com/hastexo/tutor-contrib-retirement
1. Add this repo/branch to the LMS requirements:
   `echo "git+https://github.com/openedx/openedx-event-sink-clickhouse.git@jill/retire-user#egg=openedx-event-sink-clickhouse==0.5.0" >> "$(tutor config printroot)/env/build/openedx/requirements/private.txt"`
1. Enable Aspects PII:
   `tutor config save -s ASPECTS_ENABLE_PII=True`
1. (optional) Shorten the lifetime of the user_pii dict to 10s to make testing data changes easier:
   `tutor config save -s ASPECTS_PII_CACHE_LIFETIME=10`
1. (optional) Remove the cooling off period to make it easier to process user retirements:
   `tutor config save -s RETIREMENT_COOL_OFF_DAYS=0`
1. Rebuild image:
   `tutor images build openedx --no-cache`
1. Re-init Aspects to use updated settings:
   `tutor local do init -l aspects`
1. Reboot Tutor to use new images:
   `tutor local reboot -d`

**Testing instructions:**

1. Register a new user on the platform, and update that user's Account (aka user profile) details.
    Use Superset's SQL Lab to check that the user profile change(s) flow through to the `event_sink.user_pii` table.
1. Enrol the user in a course to see external IDs created and flowing into the `event_sink.user_pii` table.
1. In the LMS Accounts page, delete the user.
1. From the command line, run the user retirement pipeline to move the user's retirement request from PENDING into COMPLETED.
    `tutor local retire-users`
1. After ASPECTS_PII_CACHE_LIFETIME elapses, check Superset SQL Lab to ensure that the user records have been deleted from `event_sink.user_pii`.

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is
      finished.
- [ ] Delete working branch (if not needed anymore)

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @pomegranited! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: event_sink_clickhouse/signals.py [hunk 1]
```diff
@@ -2,11 +2,18 @@
...
 from event_sink_clickhouse.sinks.external_id_sink import ExternalIDSInk
+from event_sink_clickhouse.sinks.user_retire import UserRetirementSink
 from event_sink_clickhouse.utils import get_model
 
+try:
+    from openedx.core.djangoapps.user_api.accounts.signals import USER_RETIREMENT_LMS_MISC
+except ImportError:
+    # Tests don't have the platform installed
+    USER_RETIREMENT_LMS_MISC = Signal()
+
```
Comment (Ian2012) [inline event_sink_clickhouse/signals.py:16]: We should try to use an openedx event for this one, I think there isn't one but I'm not sure. I can do the contribution if needed

File: tests/test_user_retire.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+from test_utils.helpers import FakeUser
+
+
+@responses.activate(  # pylint: disable=unexpected-keyword-arg,no-value-for-parameter
+    registry=OrderedRegistry
+)
+@patch("event_sink_clickhouse.sinks.user_retire.UserRetirementSink.serialize_item")
+@patch("event_sink_clickhouse.sinks.user_retire.UserRetirementSink.is_enabled")
+@patch("event_sink_clickhouse.sinks.user_retire.UserRetirementSink.get_model")
+def test_retire_user(mock_user_model, mock_is_enabled, mock_serialize_item):
```
Comment (bmtcril) [inline tests/test_user_retire.py:20]: Can we get one more test that exercises the `many` version of this, just to make sure no one breaks it in the future?

File: event_sink_clickhouse/sinks/user_retire.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class UserRetirementSink(ModelBaseSink):  # pylint: disable=abstract-method
+    """
+    Sink for user retirement events
+    """
+
+    model = "auth_user"
+    unique_key = "id"
+    clickhouse_table_name = ["user_profile", "external_id"]
```
Comment (bmtcril) [inline event_sink_clickhouse/sinks/user_retire.py:15]: It's sad that we don't have a more data driven way to deal with this list, but I think it's ok. Can you put a comment here reminding people that this list should exactly match `EVENT_SINK_CLICKHOUSE_PII_MODELS` in the Aspects plugin? We'll probably want one there, too, to try to keep these from falling out of sync.

Comment (Ian2012) [reply event_sink_clickhouse/sinks/user_retire.py:15]: And transform the PII_MODELS, and MODELS, settings into OVERRIDE_SETTINGS. So those cannot be overriden

Comment (bmtcril) [reply event_sink_clickhouse/sinks/user_retire.py:15]: I don't know if this is a desired capability, but it is possible to define other sinks in plugins, so people may want to override those settings. I'm not 100% sure about the hierarchy of settings here, but I think if there's a default setting in this app, and a setting in tutor-contrib-aspects the latter will override. Then if an operator has other needs their specific config will take precedence. I think that would be my preference.

File: tests/test_user_retire.py (hunk 2 above)
Comment (pomegranited) [reply tests/test_user_retire.py:25]: Sure thing, see https://github.com/openedx/openedx-event-sink-clickhouse/pull/63/files#diff-69e4bc0dbdddfc867532a028e3b4b6764834376c04b2ac62b5ea5e61ee928145R77

File: event_sink_clickhouse/sinks/user_retire.py (hunk 3 above)
Comment (pomegranited) [reply event_sink_clickhouse/sinks/user_retire.py]: (rebuild worked.)

**Review by Ian2012 — APPROVED** [review]

**Review by bmtcril — APPROVED** [review]
LGTM!

**Pull request merged**